### PR TITLE
Updates for SAS length

### LINKED ISSUE #20: Specify 'length' attribute of variable in XPT file
Hello,

I am writing an XPT file and I would like to specify the length attribute of the character variables. Looking at the `write.xport` function it looks like this is set as `max(c(1,nchar(var, "bytes", keepNA=FALSE)))`. Would you be open to a PR that adds the option to  pass an attribute "SASlength" that overrides the length on character variables? I've tested this out and it can be handled similarly to "SASformat".

```
mtcars2$mpg <- as.character(mtcars2$mpg)
attr(mtcars2$mpg, "SASlength") <- 20
write.xport(mtcars2, file = "test.xpt")
```
image
Thanks,
Eli

### PR DESCRIPTION
**Pull request by elimillera**
This closes #20.

I've added a `SASlength` function to set and return the `SASlength` attribute similar to the other attribute setters. I've updated the documentation, NAMESPACE, and NEWS.md files as well.

Let me know if there are any changes you want to see before sending to CRAN. R CMD CHECK is all green.

```
> library(SASxport)
> mtcars2 <- mtcars
> mtcars2$mpg <- as.character(mtcars2$mpg)
> SASlength(mtcars2$mpg) <- 20
> mtcars2$am <- as.character(mtcars2$am)
> write.xport(mtcars2, file = "test.xpt")
```

### CODE REVIEW

File: R/SASlength.R [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+`SASlength<-.default` <- function(x, value) {
+  attr(x,'SASlength') <- value
+  x
+}
+
+#' @export
+`SASlength<-.data.frame` <- function(x, value)
+{
+  if( ncol(x) != length(value) )
+    stop("vector of formats must match number of data frame columns")
```
Comment (warnes) [inline R/SASlength.R]: This should be "vector of lengths.."

File: DESCRIPTION [hunk 2]
```diff
@@ -1,7 +1,7 @@
 Package: SASxport
 Title: Read and Write 'SAS' 'XPORT' Files
-Version: 1.7.0
-Date: 2020-03-06
+Version: 1.7.1
```
Comment (warnes) [inline DESCRIPTION]: Since functionality was added, it should be 1.8.0.

File: R/write.xport.R [hunk 3]
```diff
@@ -184,13 +184,14 @@ write.xport <- function(...,
...
             varIFormat <- attr(var, "SASiformat")
+            varLength <- attr(var, "SASlength")
 
             # Convert R object to SAS object
             df[[i]] <- var <- toSAS(var, format.info=formats)
 
             # compute variable length
             if(is.character(var))
-              varLen <- max(c(1,nchar(var, "bytes", keepNA=FALSE) ) )
+              varLen <- ifelse(is.null(varLength), max(c(1,nchar(var, "bytes", keepNA=FALSE))), varLength)
```
Comment (warnes) [inline R/write.xport.R]: Split this out into several lines for readability.

File: man/write.xport.Rd [hunk 4]
```diff
@@ -86,7 +86,8 @@ write.xport(...,
   \code{\link{label}},
   \code{\link{SAStype}},
   \code{\link{SASformat}},
-  and \code{\link{SASiformat}}
+  \code{\link{SASiformat}}
```
Comment (warnes) [inline man/write.xport.Rd]: Needs a training comma.

**Review by warnes — CHANGES_REQUESTED** [review]
Add `SASlengrh` examples to the two man pages and see the other comments.

File: R/SASlength.R (hunk 1 above)
Comment (elimillera) [reply R/SASlength.R]: Updated

File: DESCRIPTION (hunk 2 above)
Comment (elimillera) [reply DESCRIPTION]: Updated

File: R/write.xport.R (hunk 3 above)
Comment (elimillera) [reply R/write.xport.R]: Split out and some comments were added for context

File: man/write.xport.Rd (hunk 4 above)
Comment (elimillera) [reply man/write.xport.Rd]: Updated

**Comment by elimillera** [comment]
@warnes Thanks for the quick review. Updates have been made.

**Comment by elimillera** [comment]
@warnes Any other changes you'd like to see on this?

File: DESCRIPTION (hunk 2 above)
Comment (warnes) [reply DESCRIPTION]: It still shows 1.7.1 instead of 1.8.0.  Did you miss a push?

File: R/write.xport.R [hunk 5]
```diff
@@ -190,10 +190,18 @@ write.xport <- function(...,
             df[[i]] <- var <- toSAS(var, format.info=formats)
 
             # compute variable length
-            if(is.character(var))
-              varLen <- ifelse(is.null(varLength), max(c(1,nchar(var, "bytes", keepNA=FALSE))), varLength)
-            else
+            if(is.character(var)){
+              # If variable is character, use varLength if available.
+              if(is.null(varLength)){
```
Comment (warnes) [inline R/write.xport.R]: Use `if(length(varLength)==0)` which catches both `NULL` and `character(0)`.

File: R/write.xport.R [hunk 6]
```diff
@@ -190,10 +190,18 @@ write.xport <- function(...,
...
+                varLen <- max(c(1,nchar(var, "bytes", keepNA=FALSE)))
+              } else {
+                varLen <- varLength
+              }
+            }
+            else{
+              # If variable is numberic, length is 8
```
Comment (warnes) [inline R/write.xport.R]: Typo:  `numberic` --> `numeric`

**Review by warnes — CHANGES_REQUESTED** [review]

**Comment by warnes** [comment]
Code looks good!  Just a couple of minor nits.

I'm going to pull down the code and run some checks now.

-G

**Comment by mstackhouse** [comment]
@warnes Any chance this could be merged soon? The length update here would allow CDISC compliant XPT file generation, which is going to be a huge benefit to the pharma industry's use of R.

**Review by gwarnes-mdsol — APPROVED** [review]

**Review by warnes — APPROVED** [review]

**Comment by elimillera** [comment]
@warnes Do you have a time frame for this being pushed up to CRAN?

**Comment by DennisSweitzer** [comment]
Good to see these updates; My team just duplicated this before seeing you fixed it.
We also found another limitation. SAS allows integers to be defined with <8 bytes of memory, but SASxport converts all integers into an 8 byte double before putting it in the buffer.
Are there any plans to fix that? I found the code in xport.numeric() that needs modification, but haven't worked with C libraries before.

**Comment by mstackhouse** [comment]
@warnes do you intend to move this version to CRAN anytime soon? It would be immensely valuable to have this PR incorporated and available to install via CRAN.

**Pull request merged**